File: apps/client-macos/Sources/TimeTrack/Projects/RefreshThrottle.swift
import Foundation

/// Rate-limits an on-demand refresh (e.g. re-fetching the project list when the menu opens) so
/// frequent triggers don't hammer the API. `shouldRefresh()` returns true at most once per
/// `minInterval`, recording the moment it allows. `clock` is injected for deterministic tests.
/// Main-thread only (the menu-open hook and the caller both run on the main thread).
final class RefreshThrottle {
    private let minInterval: TimeInterval
    private let clock: () -> Date
    private var last: Date?

    init(minInterval: TimeInterval, clock: @escaping () -> Date = Date.init) {
        self.minInterval = minInterval
        self.clock = clock
    }

    /// True if at least `minInterval` has elapsed since the last allowed refresh (records now and
    /// allows); false to skip. The first call always allows.
    func shouldRefresh() -> Bool {
        let now = clock()
        if let last, now.timeIntervalSince(last) < minInterval { return false }
        last = now
        return true
    }
}

File: apps/client-windows/src/NiftyTimer/App/RefreshThrottle.cs
namespace NiftyTimer.App;

/// <summary>
/// Rate-limits an on-demand refresh — re-fetching projects or totals as the menu opens — so a
/// person clicking the tray icon repeatedly does not hammer the API and its throttler. Ported from
/// the macOS client's <c>RefreshThrottle</c>.
///
/// <see cref="ShouldRefresh"/> allows at most once per interval and records the moment it allowed;
/// the first call always allows. Pure apart from the injected clock. UI-thread-only, like its only
/// caller.
/// </summary>
public sealed class RefreshThrottle
{
    private readonly TimeSpan _minInterval;
    private readonly Func<DateTimeOffset> _clock;
    private DateTimeOffset? _last;

    public RefreshThrottle(TimeSpan minInterval, Func<DateTimeOffset>? clock = null)
    {
        _minInterval = minInterval;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
    }

    public bool ShouldRefresh()
    {
        var now = _clock();
        if (_last is { } last && now - last < _minInterval)
        {
            return false;
        }

        _last = now;
        return true;
    }

    /// <summary>Sign-out: the next person's first menu open refreshes at once.</summary>
    public void Reset() => _last = null;
}
